rfc/001-composable-pipelines/README.md

```diff
@@ -0,0 +1,393 @@
+# Composable Pipelines
+
+- **Status**: In-Review
+- **Created**: Aug 19, 2022
+- **ETA**: Sep 19, 2022
+- **Authors**:
+  - [Matthias Vallentin](https://github.com/mavam)
+- **Contributors**:
+  - [Anthony Verez](https://github.com/netantho)
+  - [Dominik Lohmann](https://github.com/dominiklohmann)
+  - [Rémi Dettai](https://github.com/rdettai)
+  - [Thomas Peiselt](https://github.com/dispanser)
+- **Discussion**: [PR #2511](https://github.com/tenzir/vast/pull/2511)
+
+## Overview
+
+This proposal ideates a new UX around VAST's pipelines. The fundamental idea is
+that *everything is a pipeline* and build a composable UX around this concept.
+
+## Problem Statement
+
+Pipelines currently are only deployable at [predefined
+points](https://vast.io/docs/use-vast/transform). For example, they can be
+triggered for specific events on import/export, or invoked by compaction.
+
+As we are bringing pipelines to the query language, we need to solve a UX
+challenge: how do we make the data flow explicit and provide a unified
+experience, especially in the presence of custom plugins?
+
+## Solution Proposal
+
+Assume for a moment everything is a pipeline operator. Then we may write:
+
+ ```bash
+vast 'from s3://aws |
+      read json |
+      summarize count(dst) group-by src |
+      write feather |
+      to /path/to/file.feather'
+```
+
+The UX would translate seamlessly to other languages, e.g., Python:
+
+```python
+import vast as v
+await v.from("s3://aws")
+       .read_json() # or read("json")
+       .group_by("src") # could also be merged with summarize()
+       .summarize(v.count("dst"))
+       .write_feather() # or write("feather")
```

Comment:
In spark, you'd have `spark.read.avro(<path>)` for built-in formats, and `spark.read.format("delta")` for things that are not natively distributed with spark (basically plugins). So I think we should support both, offering code completion help where we can and a generic path for everything else. :+1: 

(not actionable, just a reaffirmative comment)

Reply:
looking at the `python` example below, I note that there such an expression would be possible, because you'd probably expose some intermediate pipeline builder object, and all information is freely available. The problem only surfaces in the context of a `|` that erases all previous information (based on the I/O table below).

Reply:
The same problem will probably come up in the `read` direction, when trying to read more than a single file: be it that you're interested in the file names (commonly useful to derive additional columns) or you want to read an S3 bucket full of csv files: each file possibly has a header, and a single stream of bytes does not allow to restore the file boundaries.

I think my overall concern is that `Bytes` is insufficient.

Reply:
Would my idea of `load`/`store` above fix this?